documentation/fortran-ir.md

```diff
@@ -0,0 +1,194 @@
+
+_ReturnStmt_: execution leaves the enclosing Procedure. A ReturnStmt can return and optional value. This would appear for RETURN statements or at END SUBROUTINE.
+
+_BranchStmt_: execution of the current basic block ends. If the branch is unconditional then control transfers to exactly one successor basic block. If the branch is conditional then control transfers to exactly one of two successor blocks depending on the true/false value of the condition. All successors must be in the current Procedure. Unconditional branches would appear for GOTO statements. Conditional branches would appear for IF constructs, IF statements, etc.
+
+_SwitchStmt_: Exactly one of multiple successors is selected based on the control expression. Successors are pairs of case expressions and basic blocks.  If the control expression compares to the case expression and returns true, then that control transfers to that block. There may be one special block, the default block, that is selected if none of the case expressions compares true. This would appear for SELECT CASE, SELECT TYPE, SELECT RANK, COMPUTED GOTO, WRITE with exceptional condition label specificers, alternate return specifiers, etc.
+
+_IndirectBrStmt_: A variable is loaded with the address of a basic block in the containing Procedure. Control is transferred to the contents of this variable. An IndirectBrStmt also requires a complete list of potential basic blocks that may be loaded into the variable. This would appear for ASSIGNED GOTO.
+
+Supporting ASSIGNED GOTO offers a little extra challenge as the ASSIGN GOTO statement's list of target labels is optional.  If that list is not present, then the procedure must be analyzed to find ASSIGN statements. The implementation proactively looks for ASSIGN statements and keeps a dictionary mapping an assigned Symbol to its set of targets. When constructing the CFG, ASSIGNED GOTOs can be processed as to potential targets either from the list provided in the ASSIGNED GOTO or from the analysis pass.
+
+_UnreachableStmt_: If control reaches an unreachable statement, then an error has occurred. Calls to library routines that do not return should be followed by an UnreachableStmt.  An example would be the STOP statement.
+
+#### Scope
+
+In the CFG-WP, scopes are meant to be captured by a pair of backstage statements for entering and exiting a particular scope. In structured code, these pairs would not be problematic; however, control flow in Fortran is ad hoc, particularly in legacy Fortran. In short, Fortran does not have a clean sense of structure with respect to scope.
+
+To separate concerns, FIR will construct the ad hoc CFG and impose bounding boxes over regions of that graph to demarcate and superimpose scope structures on that CFG. Any GOTO-like statements that are side-entries and side-exits to the region will be explicit.
```

Comment:
Are there actually cases in Fortran where one can jump into any part of a scope besides its starting point, apart from the special weird case of `ENTRY` statements?  You can probably simplify the problem by assuming that scopes are single-entry multi-exit regions.

Reply:
We aren't disallowing "ugly" GOTOs that jump in or out of regions to support dusty deck Fortran.

Reply:
There's extended-range DO loops, but DO loops shouldn't constitute regions unless they're OpenMP or OpenACC constructs, and those don't support extended ranges.  Other constructs with regions (`BLOCK`, `SELECT TYPE/RANK`, `ASSOCIATE`)  come with control flow constraints that we enforce (or must, if we don't).  Can you think of a specific instance of a scope where we'd need to represent an incoming flow arc that doesn't arrive at its head?